### PR TITLE
[1.16] Fix an off-by-one in mapgen's x and y filters, and add unit tests

### PR DESCRIPTION
**Pull request by stevecotton**
Fixes the currently suspected root cause of #5108, but requires further testing
to confirm that it fixes that issue.

Closes #6501 (the question of how to test the mapgen filter).

Opening the 1.16 PR first, will forward-port afterwards.
Ping @knyghtmare

### CODE REVIEW

**Comment by knyghtmare** [comment]
> Ping @knyghtmare
I have begun to run tests. I will load the WC scenario 20-30 times and see if the bug is present or not and upload the excel file here.

**Comment by knyghtmare** [comment]
My testing results. These were the occurrences I got, and the bug seems to have been fixed.
image
image

Excel File of 20 trials.
[Bug-5108-test.xlsx](https://github.com/wesnoth/wesnoth/files/7073007/Bug-5108-test.xlsx)

**Comment by CelticMinstrel** [comment]
The fix _looks_ good at a glance, but I am uncertain about one thing. Does the filter used in WC separately exclude border hexes? Because if it not, it looks like your change now means the items can be placed on the top or left border of the map. If that's not correct, the check should be changed from `>= 0` to `> 0`.

**Comment by stevecotton** [comment]
The bounds checks here are just to ensure the C++ doesn't access an invalid index into the bitset. Filtering the top and left borders hexes is done by the bitset itself.

The code in WC does exclude the borders, it filters for `"2-" .. map.width-3`.

**Comment by CelticMinstrel** [comment]
That sure is ugly though… probably we should at least change that to `map.playable_width-2` for symmetry… and I guess it's intentional, but it prevents placing them not only _on_ the border but also adjacent to the border.

**Comment by stevecotton** [comment]
Changing it to use map.playable_width is scope creep for this PR, because this PR is fixing an API bug; it doesn't need to change any of the campaign files.

Changing map.width-3 to map. playable_width-2 sounds like an off-by-one error too. Assuming the border is the same size on both sides, the difference should be even.

**Comment by CelticMinstrel** [comment]
> Changing map.width-3 to map. playable_width-2 sounds like an off-by-one error too.

Hmm… you're probably right, actually; `playable_width` should be 2 less than `width`. I guess we just have to live with the asymmetry.

**Comment by stevecotton** [comment]
I think it's ready to merge, but waiting for a review approval.

**Review by CelticMinstrel — APPROVED** [review]

**Pull request merged**
